modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableColumnHeader.java

```diff
@@ -586,8 +590,174 @@ private void doColumnAutoSize(TableColumnBase<?,?> column, int cellsToMeasure) {
 
         // if the prefWidth has been set, we do _not_ autosize columns
         if (prefWidth == DEFAULT_COLUMN_WIDTH) {
-            TableSkinUtils.resizeColumnToFitContent(getTableSkin(), column, cellsToMeasure);
-//            getTableViewSkin().resizeColumnToFitContent(column, cellsToMeasure);
+            resizeColumnToFitContent(column, cellsToMeasure);
+        }
+    }
+
+    /**
+     * Resizes the given column based on the preferred width of all items contained in it. This can be potentially very
+     * expensive if the number of rows is large. Subclass can either call this method or override it (no need to call
+     * {@code super()}) to provide their custom algorithm.
+     *
+     * @param tc      the column to resize
+     * @param maxRows the number of rows considered when resizing. If -1 is given, all rows are considered.
+     * @since 12
+     */
+    protected void resizeColumnToFitContent(TableColumnBase<?, ?> tc, int maxRows) {
+        if (!tc.isResizable()) return;
```

Comment:
why do we need the column as parameter? Now that the header is the responsible party, it should be concerned about its own column only - or not?

Reply:
In the `rectMousePressed` event of `NestedTableColumnHeader`, the column is retrieved like that :
```
Rectangle rect = (Rectangle) me.getSource();
        TableColumnBase column = (TableColumnBase) rect.getProperties().get(TABLE_COLUMN_KEY);

```
I do not know if it's guaranteed that the column retrieved is the TableColumn associated with the current NestedTableColumnHeader. If yes, we can indeed remove the TableColumn parameter.

Reply:
good point, I'll check later

Reply:
with the current impl we need the tableColumn as parameter: the header that's stored in the dragRect properties is the parent of the tableColumnHeader that represents it. 

But then, we are back to letting unrelated parties do the sizing: the parent does the sizing on behalf of the real header: pre-9 it was the skin, since then it was the utils, now it's the parent. Ideally, the sizing would be handled by the "real" header, IMO. That's where the current sizing code resides (in TableColumnHeader) but it is invoked from upper with an arbitrary column (which is a child of the column the nested is responsible for) ... hmm ...

Reply:
To clarify my base a bit: the setup of auto-sizing in core is a unusual (read: never was/is what I expected ;)- it's managed by nestedColumnHeader when it should be handled by columnHeader and then only on its own tableColumn (not others) and no other party should interfere with that sizing. The mouse interaction is installed in Nested because it manages the dragRects and the handlers are intalled on those. That's fine, iff the handler delegates the actual sizing to the header that's representing the column. 

In the pressed handler, something like:

````java

TableColumnBase column = (TableColumnBase) rect.getProperties().get(TABLE_COLUMN_KEY);
 NestedTableColumnHeader parentHeader = (NestedTableColumnHeader) rect.getProperties().get(TABLE_COLUMN_HEADER_KEY);
   ...

if (me.getClickCount() == 2 && me.isPrimaryButtonDown()) {
     // the user wants to resize the column such that its
     // width is equal to the widest element in the column
     TableHeaderRow tableHeader = parentHeader.getTableHeaderRow(); 
     TableColumnHeader columnHeader = tableHeader.getColumnHeaderFor(column);
     // note the missing column param ... 
     columnHeader.resizeColumnToFitContent(-1);
 } else {
  ....
}
````

Just a snippet, untested and not fully thought through :) In particular, didn't check other locations that message the auto-size method.